README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->

# OutliersO3

OutliersO3 is for visualising results of outlier analyses. Overview of
Outliers (O3) plots show which cases are identified as potential
outliers for different combinations of variables from a dataset.

You can compare sets of outliers identified by up to six different
methods. You can also compare results for a single method at up to three
different tolerance
levels.

## Install in the usual way

install.packages(“OutliersO3”)

## What outliers are there amongst the genuine banknotes in the Swiss banknote dataset?

Flury and Riedwyl introduced the famous banknote dataset in their
excellent book on multivariate statistics. There are measurements on 100
genuine banknotes and on 100 counterfeit banknotes. Presumably the
genuine notes should all be very similar.

The method mvBACON from *robustX* has been used to identify possible
outliers. There are 6 numeric measurements of the notes, so there are 63
possible variable combinations. An O3 plot has one row for each variable
combination for which outliers were found and those variables are
specified by the relevant columns on the left of the plot. The cases
identified as outliers for at least one combination each get a column to
the right of the plot.

``` r
library(OutliersO3)
#> Registered S3 method overwritten by 'GGally':
#>   method from   
#>   +.gg   ggplot2
data(banknote, package="mclust")
data <- banknote %>% filter(Status=="genuine") %>% select(-Status)
pB <- O3prep(data, method="BAC", tols=c(0.05, 0.01, 0.001), boxplotLimits=c(6,10,12))
pX <- O3plotT(pB)
#> Warning: Vectorized input to `element_text()` is not officially supported.
#> Results may be unexpected or may change in future versions of ggplot2.
pX$gO3
```

![](man/figures/README-unnamed-chunk-3-1.png)<!-- -->

The O3 plot shows outliers found by the mvBACON method for three
tolerance levels. Two banknotes, X71 and X5, are only identified for a
few combinations mainly at a level of 0.05. A further banknote, X40 is
identified more often, mostly at a level of 0.01. Two banknotes, X1 and
X70, were identified as outliers at a level of 0.001 for the combination
of attributes Length and Right and for Diagonal on its own respectively.
When X1 is identified as an outlier at other levels the attribute Right
is always involved. The supporting parallel coordinate plot suggests
why:

``` r
pX$gpcp
```

![](man/figures/README-unnamed-chunk-4-1.png)<!-- -->

This plot also suggests that all five cases identified as potential
outliers are relatively extreme on at least one of the six attributes.

There are more examples in the package vignettes.


NEWS
Outliers03 v0.6.3 (Release date: 2020-04-27)
==============
Changed how labels are drawn because "Vectorized input to `element_text()` is not officially supported" by ggplot2.
Some small changes to help files on the tolerance parameters and on plot margins (thanks to Jan Greyling for suggesting this).

Outliers03 v0.6.2 (Release date: 2019-03-27)
==============
Reference added to JCGS article on O3 plots (in press).
Dataset Election2005 included in package, so that \pkg{mbgraphic} is no longer suggested.
Some small additions to help files.

Outliers03 v0.6.1 (Release date: 2019-02-27)
==============
The authors of the package cellWise have renamed their function DetectDeviatingCells DDC and this required a minor code change in OutliersO3 and amendments to help files.

Outliers03 v0.6 (Release date: 2019-01-21)
==============
(Thanks to anonymous referees and not-so-anonymous JCGS editors for suggesting some of these changes.)

Use mvBacon and covMcd for their respective 1-d outliers and not boxplot limits.

Changed the chi-square degrees of freedom used in setting thresholds for covMcd to number of variables in subspace from number of variables in dataset.

Added legends to O3 plots and no longer include plot titles.

There are now two white columns separating the left and right blocks of O3 plots.

Added the options sortVars (whether variables in an O3 plot are sorted or not) and coltxtsize (text size of column names) to O3plotT and O3plotM.

Simplified entry for the boxplotLimits parameter vector.


Outliers03 v0.5.4 (Release date: 2018-02-08)
==============
Corrected the index entries for the vignettes so that they are all different.


Outliers03 v0.5.3 (Release date: 2018-02-06)
==============
Relaxed the dependence on R from ≥ R 3.4.0 to ≥ R 3.3.0 to ensure that the package would run under r-oldrel-osx-x86_64 and thus pass all CRAN package checks.


Outliers03 v0.5.1 (Release date: 2018-02-05)
==============
There are three major changes to the package.

1) The main function O3plot has been split into three parts.  O3prep sets up a list of lists of outlier indices and distances to be plotted by O3plotM (if more than one outlier method is involved) or by O3plotT (if the plot is for one method and up to 3 different tolerance levels).  Separating preparation and plotting means that users can prepare outlier results with their own methods and code to then draw and O3 plot with this package.

2) If an O3 plot is to be drawn for more than one method, then the outlier tolerance levels can be set individually for each method.  This is now the default as using the same levels for all methods was rarely satisfactory.  A vignette has been added to illustrate this.

3) Argument names have been "R-standardised" and internal functions hidden.  (Thanks to Michael Friendly for helpful comments and to Bill Venables for sage advice.)

The output now includes a table of all outliers found by case, variable combination, and method/tolerance level.  It also includes a three-dimensional array of all the outlier distances/scores by variable combination by method for all methods that provide them.

Two further plots have been added, an O3 plot for more than two methods in which only case combinations identified by at least two outliers are displayed, and a parallel coordinate plot of outlier distances (when provided by methods) either by methods (O3plotM) or by combinations (O3plotT).

The method covMcd{robustbase} has been added as an optional method.

For an O3 plot with one method and more than one outlier level, you can enter the levels in the arguments tols and boxplotLimits in any order. (Thanks to Tae-Rae Kim for this suggestion.)

For an O3 plot with more than one method, you can enter the methods in any order.

If a dataset includes case identifiers, then these can be used as labels for the case columns in the O3 plot through the argument caseNames.

O3 plot rows are now sorted by numbers of outliers within size of variable combinations. (Thanks to Nina Wu for pointing out that this was not working as intended.)

The default colours in an O3 plot for comparing two methods have been changed to be more distinctive.

Colours can be specified using either colours or colors (or col) in O3plotColours.


Outliers03 v0.2.1 (Release date: 2017-09-26)
==============

First version on CRAN.


R/O3plotM.R
# quiets concerns of R CMD check
if(getRversion() >= "2.15.1")  utils::globalVariables(c("ID",
            "outlierIndices", "pID", "psN", "psNx", "sB", "s1B",
            "sN", "sumR", "sumS", "sumV", "xsumR", "oh", "HDo", "case", "Case", "X", "SsB"))

# Main function--------------

O3plotM <- function(outResults, caseNames=paste0("X", 1:nrow(outResults$data)),
                   sortVars=TRUE, coltxtsize=14, O3control=O3plotColours()) {

ouF <- outResults$data
n1 <- ncol(ouF)
n2 <- nrow(ouF)
nw <- outResults$nw
mm <- outResults$mm
mx <- length(mm)
if(mx < 2) {
stop("You should use O3PlotT for plotting results for a single method", call.=FALSE)
    }
tols <- outResults$tols
outList <- outResults$outList

# There are mx outM's and l2a gives the indices of cases that are ever outliers
  l1a <- vector("list", mx)
  l2a <- NULL
  for  (j in 1:mx) {
    l1a[[j]] <- rlist::list.cases(outList[[j]], outlierIndices)
    l2a <- union(l2a, l1a[[j]])
    }

# Stop if no outliers are found--------------
  nz <- length(l2a)
  if (nz < 1) {
    cat("No outliers found")
    }
  if (nz > 0) {
    #Set up an array with cases ever found to be outliers and one with outlier scores
    zz <- rep(0, nw*nz*mx)
    dim(zz) <- c(nw, nz, mx)
    colnames(zz) <- paste0("X", l2a)
    Cs <- rep(0, mx*nw*n2)
    dim(Cs) <- c(n2, nw, mx)

# Fill a matrix with outlier indices and one with outlier scores for each method
        for (j in 1:mx) {
          l1 <- rlist::list.ungroup(outList[[j]])
          #Prepare a comparison plot if there are only two methods
             for (k in 1:nw) {
               if (mx == 2) {
                 if (length(l1[[3*k-1]]) > 0) {
              zz[k, paste0("X", l1[[3*k-1]]), j] <- 3+j/100
              }
          } else {
          if (mx > 2) {
              if (length(l1[[3*k-1]]) > 0) {zz[k, paste0("X", l1[[3*k-1]]), j] <- 7}
              }
            }
          Cs[, k, j] <- l1[[3*k]] 
          }
        }

        zs <- apply(zz, c(1,2), sum)

# Numbers of outliers found by each method
        zt <- apply(zz, c(2,3), sum)
        zt[zt>0] <- 1
        nOut <- colSums(zt)
        names(nOut) <- mm

# Adjust scores so that outliers found by all methods are red if more than 2
        if (mx > 2) {
          zs[zs>0] <- zs[zs>0] + 7*(6-mx)
        }
      }

#Add variable columns and gap columns to matrix--------------
    zv <- rep(0, nw*(n1 + 2))
    dim(zv) <- c(nw, n1+2)
    colnames(zv) <- c(names(ouF), "gap1", "gap2")
    for (k in 1:nw) {
      zv[k, l1[[3*k-2]]] <- 1/1000
      zv[k, (n1+1):(n1+2)] <- 2/1000
    }
    colnames(zs) <- caseNames[l2a]
    z1 <- data.frame(cbind(zv, zs))

# Remove combinations with no outliers--------------
    if (nz > 1) {
    z1 <- z1[rowSums(z1[ , (n1+3):(n1+2+nz)]) > 0, ]
    } else {
      if (nz == 1)
      z1 <- z1[z1[ , (n1+3)] > 0, ]
    }

# Sort and prepare----------------------------
    z1p <- sortO3(z1, n1=n1, nz=nz, ouF, sortVars)

# Plot------------------------------------------

# Prepare blue lines separating numbers of combinations--------------
    nc <- nrow(z1)
    tx <- rep(0,n1)
    t1d <- data.frame(table((nc+1)*rowSums(z1[ ,1:n1])))
    tx[t1d$Var1] <- t1d$Freq
    ty <- cumsum(rev(tx)) + 0.5
    tw <- ty[ty > 0.5 & ty < nc + 0.5]

# Set the colours for the plot------------------
# First add the background colours
    outCols <- c("grey95", "grey60", "white", O3control$colours)

    z1p <- z1p %>% mutate(s1B = factor(sB, levels = c("0", "0.001", "0.002", "3",
      "4", "5", "3.01", "3.02", "6.03", "7", "14", "21", "28", "35", "42")))
    names(outCols) <- levels(z1p$s1B)
    levels(z1p$psNx)[n1+1] <- " "
    levels(z1p$psNx)[n1+2] <- "  "
    
    ga <- ggplot(z1p, aes(psNx, forcats::fct_rev(pID), group = s1B, fill = s1B)) +
      geom_tile(colour = "grey30") + labs(x = NULL, y = NULL) +
      theme(legend.position = "bottom", panel.background = element_rect(fill = "white"),
      axis.text.x = element_text(size=coltxtsize, angle=45, hjust=0, vjust=0),
      axis.ticks.y = element_blank(), axis.text.y = element_blank()) +
      scale_x_discrete(position = "top") +
      geom_hline(yintercept = tw, lty = 3, colour = "blue", size=0.75)
      if (mx == 2) {
         gO3 <- ga  + scale_fill_manual(name = "Outliers identified by", values = outCols,
      breaks=c("3.01", "3.02", "6.03"), labels=c(mm[1], mm[2], "both"))
         } else {
      if (mx == 3) {
      gO3 <- ga + scale_fill_manual(name = "No of methods identifying outliers",
      values = outCols, breaks=c("28", "35", "42"), labels=c("1", "2", "3"))
      } else {
      if (mx == 4) {
      gO3 <- ga + scale_fill_manual(name = "No of methods identifying outliers",
      values = outCols, breaks=c("21", "28", "35", "42"), labels=c("1", "2", "3", "4"))
      } else {
      if (mx == 5) {
      gO3 <- ga + scale_fill_manual(name = "No of methods identifying outliers",
      values = outCols, breaks=c("14", "21", "28", "35", "42"),
      labels=c("1", "2", "3", "4", "5"))
      } else {
           gO3 <- ga  + scale_fill_manual(name = "No of methods identifying outliers",
           values = outCols, breaks=c("7", "14", "21", "28", "35", "42"),
           labels=c("1", "2", "3", "4", "5", "6"))
             }
           }
         }
       }
       
# Draw an O3 plot for 3 or more methods excluding outliers which are only found by 1 method
      if (mx > 2) {
            z1q <- z1p
            z1q[z1q$sB==7*(7-mx), "sB"] <- 0
            z1q <- z1q %>% group_by(psNx) %>% mutate(SsB = sum(sB)) %>% filter(SsB > 0)
            z1q <- z1q %>% mutate(s1B = factor(sB, levels = c("0", "0.001", "0.002", "3",
            "4", "5", "3.01", "3.02", "6.03", "7", "14", "21", "28", "35", "42")))
            names(outCols) <- levels(z1q$s1B)

      gb <- ggplot(z1q, aes(psNx, forcats::fct_rev(pID), group = s1B, fill = s1B)) +
        geom_tile(colour = "grey30") + labs(x = NULL, y = NULL) +
        scale_fill_manual(name = "s1B", values = outCols) +
        theme(plot.title = element_text(size = 18, hjust = 0.5),
        legend.position = "none", panel.background = element_rect(fill = "white"),
        axis.text.x = element_text(size=coltxtsize, angle=45, hjust=0, vjust=0),
        axis.ticks.y = element_blank(), axis.text.y = element_blank()) +
        scale_x_discrete(position = "top") +
        geom_hline(yintercept = tw, lty = 3, colour = "blue")
        gO3x <- gb  + ggtitle(paste("O3 plot of outliers found by at least 2 of",
                mx, "methods"))
                 } else {
                 gO3x <- NULL
                 }

# Set up PCP with cases ever identified as outliers coloured red--------------
    ouF1 <- ouF
# Create highlighting variable
    ouF1$oh <- rep(0, nrow(ouF))
# Outliers are listed in l2a
    ouF1[l2a, "oh"] <- "A"
    ouF1 <- ouF1 %>% mutate(alev = ifelse(oh == 0, 0.5, 1))
    gp <- ggparcoord(ouF1 %>% arrange(oh), scale = "uniminmax", columns=1:n1,
      groupColumn="oh", alphaLines="alev") + labs(x = NULL, y = NULL) +
      scale_colour_manual(values = c("grey70", "red")) +
      theme(plot.title = element_text(size = 18, hjust = 0.5),
      legend.position = "none", axis.ticks.y = element_blank(),
      axis.text.y = element_blank())
    gpcp <- gp  + ggtitle(paste("Cases ever found to be outliers by",
        mx, "methods"))

# Create dataset of outliers by case, combination and method--------------
        gg <- data.frame(matrix(zv[,1:n1], nrow = nw))
        if(nw > 1) {
          colnames(gg) <- colnames(zv[,1:n1])
        } else {
          colnames(gg) <- names(zv[,1:n1])
        }
        gg[gg>0] <- 1
        gg <- unite(gg, "Combination", c(1:n1), sep="")
        rownames(zz) <- paste0("c", gg$Combination)
        colnames(zz) <- caseNames[l2a]
        dimnames(zz)[[3]] <- names(nOut)
        zz[zz>0] <- 1
        outs <- memisc::to.data.frame(zz, as.vars = 0, name = "Outlier")
        colnames(outs)[1:3] <- c("Combination", "Name", "Method")
        outs <- outs[outs$Outlier > 0, 1:3]
        Cases <- data.frame(caseNames, Case=1:nrow(outResults$data))
        outy <- merge(outs, Cases, by.x = "Name", by.y = "caseNames")

# Draw a pcp of method distances (scores) for highest combination with outliers highlighted
        if(nw > 1) {
          dimnames(Cs)[[2]] <- rownames(zz)
        } else {
          dimnames(Cs)[[2]] <- list(rownames(zz))
        }
        dimnames(Cs)[[3]] <- mm
        Cx <- data.frame(Cs[ , nw, ])
        Cx$oh <- rep(0, n2)
        Cx[l2a, "oh"] <- "A"
        colnames(Cx) <- c(mm, "oh")
        if("HDo" %in% mm) {
        Cx <- Cx %>% select(-HDo) #HDo provides no distances
        }
        mx <- length(mm)-1*("HDo" %in% mm)
        if(mx > 1) {
        gc <- ggparcoord(Cx %>% arrange(oh), scale = "uniminmax", columns=1:mx,
          groupColumn="oh") + labs(x = NULL, y = NULL) +
          scale_colour_manual(values = c("grey70", "red")) +
          theme(plot.title = element_text(size = 18, hjust = 0.5),
          legend.position = "none", axis.ticks.y = element_blank(),
          axis.text.y = element_blank())
        gMethods <- gc  + ggtitle(paste("Distances for each method with outliers highlighted"))
        } else {
        gMethods <- NULL}
    return(list(nOut = nOut, gpcp = gpcp, gO3 = gO3, gO3x=gO3x, gMethods = gMethods, outsTable = outy, Cs = Cs))
  }
